### PR TITLE
e2e: add scaling case for TidbCluster

### PR DESCRIPTION
**Pull request by dragonly**
### What problem does this PR solve?

Add scaling case for TidbCluster.

ref #3652

### What is changed and how does it work?

Add scaling case for TidbCluster.

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [x] E2E test 
- [ ] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
None
```

### CODE REVIEW

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 1]
```diff
@@ -1487,36 +1491,7 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
 	})
 
 	ginkgo.Context("when stores number is equal to 3", func() {
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1493]: Should this be removed too?

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 2]
```diff
@@ -1616,6 +1591,198 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
 		err = oa.WaitForTidbClusterReady(tc, 10*time.Minute, 10*time.Second)
 		framework.ExpectNoError(err, "failed to wait for TidbCluster %s/%s components ready", tc.Namespace, tc.Name)
 	})
+
+	ginkgo.Context("Scale in", func() {
+		ginkgo.Context("and then scale out", func() {
+			ginkgo.It("should work for PD", func() {
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1597]: Please also add cases for TiKV, TiDB, TiFlash, TiCDC, Pump.

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 3]
```diff
@@ -1616,6 +1591,198 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+							if !ok {
+								log.Logf("PVC %s/%s does not have annotation tidb.pingcap.com/pvc-defer-deleting", pvc.GetNamespace(), pvc.GetName())
+								return false, nil
+							}
+							pvcUIDs[pvc.Name] = string(pvc.UID)
+						}
+					}
+					return true, nil
+				})
+				framework.ExpectNoError(err, "expect PVCs of scaled in Pods to have label tidb.pingcap.com/pvc-defer-deleting")
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1641]:
```suggestion
				framework.ExpectNoError(err, "expect PVCs of scaled in Pods to have annotation tidb.pingcap.com/pvc-defer-deleting")
```

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 4]
```diff
@@ -1616,6 +1591,198 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+					FieldSelector: fmt.Sprintf("involvedObject.name=%s,involvedObject.kind=TidbCluster", tc.Name),
+				}
+				event, err := c.CoreV1().Events(ns).List(options)
+				if err != nil {
+					log.Logf("failed to list events with options: +%v", options)
+					return false, nil
+				}
+				for _, event := range event.Items {
+					log.Logf("found event: %+v", event)
+					if event.Reason == "FailedScaleIn" {
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1746]: How can we make sure that this event is for **PD** in **this TC**?

File: tests/e2e/tidbcluster/tidbcluster.go (hunk 1 above)
Comment (dragonly) [reply tests/e2e/tidbcluster/tidbcluster.go:1493]: removed

File: tests/e2e/tidbcluster/tidbcluster.go (hunk 4 above)
Comment (dragonly) [reply tests/e2e/tidbcluster/tidbcluster.go:1746]: added message matching with "PD" for now

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 5]
```diff
@@ -1616,6 +1587,245 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+					log.Logf(fmt.Sprintf("%s is in ScalePhase", comp))
+					ginkgo.By("Wait for tc ready")
+					err = oa.WaitForTidbClusterReady(tc, 3*time.Minute, 10*time.Second)
+					framework.ExpectNoError(err, "failed to wait for TidbCluster %s/%s ready after scale in pd", ns, tc.Name)
+					log.Logf("tc is ready")
+
+					pvcUIDs := make(map[string]string)
+					ginkgo.By("Check PVC annotation tidb.pingcap.com/pvc-defer-deleting")
+					err = wait.Poll(10*time.Second, 3*time.Minute, func() (done bool, err error) {
+						for ordinal := 3; ordinal < 5; ordinal++ {
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1639]: The ordinal is different for different components.

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 6]
```diff
@@ -1616,6 +1587,245 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+								pvcSelector, err = member.GetPVCSelectorForPod(tc, v1alpha1.PDMemberType, int32(ordinal))
+							} else if comp == "TiKV" {
+								pvcSelector, err = member.GetPVCSelectorForPod(tc, v1alpha1.TiKVMemberType, int32(ordinal))
+							} else if comp == "TiDB" {
+								pvcSelector, err = member.GetPVCSelectorForPod(tc, v1alpha1.TiDBMemberType, int32(ordinal))
+							}
+							framework.ExpectNoError(err, "failed to get PVC selector for tc %s/%s", tc.GetNamespace(), tc.GetName())
+							pvcs, err := c.CoreV1().PersistentVolumeClaims(ns).List(metav1.ListOptions{LabelSelector: pvcSelector.String()})
+							framework.ExpectNoError(err, "failed to list PVCs with selector: %v", pvcSelector)
+							for _, pvc := range pvcs.Items {
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1651]: We also need to check `len(pvcs)==0`.

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 7]
```diff
@@ -1616,6 +1587,245 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+					} else if comp == "TiDB" {
+						utiltc.MustWaitForTiDBPhase(cli, tc, v1alpha1.ScalePhase, 3*time.Minute, 10*time.Second)
+					}
+					log.Logf(fmt.Sprintf("%s is in ScalePhase", comp))
+					ginkgo.By("Wait for tc ready")
+					err = oa.WaitForTidbClusterReady(tc, 3*time.Minute, 10*time.Second)
+					framework.ExpectNoError(err, "failed to wait for TidbCluster %s/%s ready after scale out %s", ns, tc.Name, comp)
+
+					ginkgo.By(fmt.Sprintf("Check PVCs are recreated for newly scaled out %s", comp))
+					for ordinal := 3; ordinal < 5; ordinal++ {
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1693]: ditto.

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 8]
```diff
@@ -1616,6 +1587,245 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+							framework.ExpectEqual(ok, false, "expect PVC %s/%s not to have annotation tidb.pingcap.com/pvc-defer-deleting", pvc.GetNamespace(), pvc.GetName())
+							pvcUIDString := pvcUIDs[pvc.Name]
+							framework.ExpectNotEqual(string(pvc.UID), pvcUIDString)
+						}
+					}
+				})
+			}
+		})
+
+		ginkgo.Context("while concurrently upgrade should work for", func() {
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1718]: Please also help cover the TiDB and TiKV.

File: tests/e2e/util/tidbcluster/tidbcluster.go [hunk 9]
```diff
@@ -74,6 +74,45 @@ func WaitForTidbClusterReady(c versioned.Interface, ns, name string, timeout tim
 	})
 }
 
+func MustWaitForPDPhase(c versioned.Interface, tc *v1alpha1.TidbCluster, phase v1alpha1.MemberPhase, timeout, pollInterval time.Duration) {
```
Comment (DanielZhangQD) [inline tests/e2e/util/tidbcluster/tidbcluster.go:77]: Please combine the three functions into one function.

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 10]
```diff
@@ -1616,6 +1587,245 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+				ginkgo.It(fmt.Sprintf("should work for %s", comp), func() {
+					ginkgo.By("Deploy initial tc")
+					tc := fixture.GetTidbCluster(ns, fmt.Sprintf("scale-out-scale-in-%s", strings.ToLower(comp)), utilimage.TiDBV4)
+					if comp == "PD" {
+						tc.Spec.PD.Replicas = 5
+					} else if comp == "TiKV" {
+						tc.Spec.TiKV.Replicas = 4
+					} else if comp == "TiDB" {
+						tc.Spec.TiDB.Replicas = 3
+					}
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1607]: Update to `switch/case`?

File: tests/e2e/tidbcluster/tidbcluster.go (hunk 6 above)
Comment (dragonly) [reply tests/e2e/tidbcluster/tidbcluster.go:1651]: check for what

File: tests/e2e/tidbcluster/tidbcluster.go (hunk 8 above)
Comment (dragonly) [reply tests/e2e/tidbcluster/tidbcluster.go:1737]: covered

File: tests/e2e/tidbcluster/tidbcluster.go (hunk 6 above)
Comment (DanielZhangQD) [reply tests/e2e/tidbcluster/tidbcluster.go:1667]: What if `len(pvcs)==0` for PD/TiKV?

Comment (dragonly) [reply tests/e2e/tidbcluster/tidbcluster.go:1670]: fixed

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 11]
```diff
@@ -1607,6 +1578,327 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+				return nil
+			})
+			framework.ExpectNoError(err, "failed to scale in tikv")
+
+			ginkgo.By("Expect up stores number stays 3")
+			pdClient, cancel, err := proxiedpdclient.NewProxiedPDClient(c, fw, ns, tc.Name, false)
+			framework.ExpectNoError(err, "create pdClient error")
+			defer cancel()
+			storesInfo, err := pdClient.GetStores()
+			framework.ExpectNoError(err, "get stores info error")
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1890]: Should we move these into the `wait` block below?

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 12]
```diff
@@ -1607,6 +1578,327 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+
+			_ = wait.PollImmediate(5*time.Second, 3*time.Minute, func() (bool, error) {
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1892]:
```suggestion
			_ = wait.PollImmediate(15*time.Second, 3*time.Minute, func() (bool, error) {
```

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 13]
```diff
@@ -1607,6 +1578,327 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+						utiltc.MustWaitForComponentPhase(cli, tc, v1alpha1.PDMemberType, v1alpha1.ScalePhase, 3*time.Minute, 10*time.Second)
+					case v1alpha1.TiKVMemberType:
+						utiltc.MustWaitForComponentPhase(cli, tc, v1alpha1.TiKVMemberType, v1alpha1.ScalePhase, 3*time.Minute, 10*time.Second)
+					case v1alpha1.TiDBMemberType:
+						utiltc.MustWaitForComponentPhase(cli, tc, v1alpha1.TiDBMemberType, v1alpha1.ScalePhase, 3*time.Minute, 10*time.Second)
+					}
+					log.Logf(fmt.Sprintf("%s is in ScalePhase", comp))
+					ginkgo.By("Wait for tc ready")
+					err = oa.WaitForTidbClusterReady(tc, 3*time.Minute, 10*time.Second)
+					framework.ExpectNoError(err, "failed to wait for TidbCluster %s/%s ready after scale in pd", ns, tc.Name)
```
[the remark this answers is not in the document]
Comment (dragonly) [reply tests/e2e/tidbcluster/tidbcluster.go:1638]: fixed

File: tests/e2e/tidbcluster/tidbcluster.go (hunk 11 above)
Comment (dragonly) [reply tests/e2e/tidbcluster/tidbcluster.go:1890]: fixed

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 14]
```diff
@@ -1966,6 +1937,327 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+						for ordinal := replicasSmall; ordinal < replicasLarge; ordinal++ {
+							var pvcSelector labels.Selector
+							switch comp {
+							case v1alpha1.PDMemberType:
+								pvcSelector, err = member.GetPVCSelectorForPod(tc, v1alpha1.PDMemberType, int32(ordinal))
+							case v1alpha1.TiKVMemberType:
+								pvcSelector, err = member.GetPVCSelectorForPod(tc, v1alpha1.TiKVMemberType, int32(ordinal))
+							case v1alpha1.TiDBMemberType:
+								pvcSelector, err = member.GetPVCSelectorForPod(tc, v1alpha1.TiDBMemberType, int32(ordinal))
+							}
```
Comment (csuzhangxc) [inline tests/e2e/tidbcluster/tidbcluster.go:2012]: Can we just use `pvcSelector, err = member.GetPVCSelectorForPod(tc, comp, int32(ordinal))` here?

if so, ditto for L2066~2072.

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 15]
```diff
@@ -1966,6 +1937,327 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+							pvcs, err := c.CoreV1().PersistentVolumeClaims(ns).List(metav1.ListOptions{LabelSelector: pvcSelector.String()})
+							framework.ExpectNoError(err, "failed to list PVCs with selector: %v", pvcSelector)
+							if comp == v1alpha1.PDMemberType || comp == v1alpha1.TiKVMemberType {
+								framework.ExpectNotEqual(len(pvcs.Items), 0, "expect at least one PVCs for %s", comp)
+							}
+							for _, pvc := range pvcs.Items {
+								annotations := pvc.GetObjectMeta().GetAnnotations()
+								log.Logf("pvc annotations: %+v", annotations)
+								_, ok := annotations["tidb.pingcap.com/pvc-defer-deleting"]
+								// framework.ExpectEqual(ok, true, "expect PVC %s/%s to have annotation tidb.pingcap.com/pvc-defer-deleting", pvc.GetNamespace(), pvc.GetName())
```
Comment (csuzhangxc) [inline tests/e2e/tidbcluster/tidbcluster.go:2023]: remove this?

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 16]
```diff
@@ -1966,6 +1937,327 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+					framework.ExpectNoError(err, "failed to scale in %s for TidbCluster %s/%s", comp, ns, tc.Name)
+					ginkgo.By(fmt.Sprintf("Wait for %s to be in ScalePhase", comp))
+					switch comp {
+					case v1alpha1.PDMemberType:
+						utiltc.MustWaitForComponentPhase(cli, tc, v1alpha1.PDMemberType, v1alpha1.ScalePhase, 3*time.Minute, 10*time.Second)
+					case v1alpha1.TiKVMemberType:
+						utiltc.MustWaitForComponentPhase(cli, tc, v1alpha1.TiKVMemberType, v1alpha1.ScalePhase, 3*time.Minute, 10*time.Second)
+					case v1alpha1.TiDBMemberType:
+						utiltc.MustWaitForComponentPhase(cli, tc, v1alpha1.TiDBMemberType, v1alpha1.ScalePhase, 3*time.Minute, 10*time.Second)
+					}
```
Comment (csuzhangxc) [inline tests/e2e/tidbcluster/tidbcluster.go:1993]: Can we use `utiltc.MustWaitForComponentPhase(cli, tc, comp, v1alpha1.ScalePhase, 3*time.Minute, 10*time.Second)` here?

if so, ditto for L2049 ~ L2056.

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 17]
```diff
@@ -1966,6 +1937,327 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+						case v1alpha1.PDMemberType:
+							tc.Spec.PD.Version = pointer.StringPtr(utilimage.TiDBV4)
+						case v1alpha1.TiKVMemberType:
+							tc.Spec.TiKV.Version = pointer.StringPtr(utilimage.TiDBV4)
+						case v1alpha1.TiDBMemberType:
+							tc.Spec.TiDB.Version = pointer.StringPtr(utilimage.TiDBV4)
+						}
+						return nil
+					})
+					framework.ExpectNoError(err, "failed to scale in %s for TidbCluster %s/%s", comp, ns, tc.Name)
```
Comment (csuzhangxc) [inline tests/e2e/tidbcluster/tidbcluster.go:2150]: please fix this message and the one in L2154.

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 18]
```diff
@@ -1966,6 +1937,327 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+						wrongImage := true
+						for _, c := range pod.Spec.Containers {
+							log.Logf("container image: %s", c.Image)
+							if fmt.Sprintf("pingcap/%s:%s", comp, utilimage.TiDBV4) == c.Image {
+								wrongImage = false
+								break
+							}
+						}
+						if wrongImage {
+							log.Failf("%s Pod has wrong image, expected %s", comp, utilimage.TiDBV4)
```
Comment (csuzhangxc) [inline tests/e2e/tidbcluster/tidbcluster.go:2183]: How about using `framework.Failf` instead?

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 19]
```diff
@@ -1966,6 +1937,327 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+			framework.ExpectNoError(err, "failed to wait for FailedScaleIn event")
+		})
+
+		ginkgo.It("TiKV from >=3 replicas to <3 should be forbidden", func() {
+			ginkgo.By("Deploy initial tc")
+			tc := fixture.GetTidbCluster(ns, "scale-in-tikv", utilimage.TiDBV4)
+			tc, err := cli.PingcapV1alpha1().TidbClusters(tc.Namespace).Create(tc)
+			framework.ExpectNoError(err, "Expected create tidbcluster")
+			err = oa.WaitForTidbClusterReady(tc, 30*time.Minute, 5*time.Second)
+			framework.ExpectNoError(err, "Expected get tidbcluster")
```
[the remark this answers is not in the document]
Comment (dragonly) [reply tests/e2e/tidbcluster/tidbcluster.go:2235]: done

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 20]
```diff
@@ -1966,6 +1937,327 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+				return nil
+			})
+			framework.ExpectNoError(err, "failed to scale in tikv")
+
+			ginkgo.By("Expect up stores number stays 3")
+			pdClient, cancel, err := proxiedpdclient.NewProxiedPDClient(c, fw, ns, tc.Name, false)
+			framework.ExpectNoError(err, "create pdClient error")
+			defer cancel()
+
+			_ = wait.PollImmediate(15*time.Second, 3*time.Minute, func() (bool, error) {
```
[the remark this answers is not in the document]
Comment (dragonly) [reply tests/e2e/tidbcluster/tidbcluster.go:2249]: Sorry I didn't catch... What do you suggest to change?

File: tests/e2e/tidbcluster/tidbcluster.go (hunk 18 above)
Comment (dragonly) [reply tests/e2e/tidbcluster/tidbcluster.go:2183]: They're actually different wrappers for `FailfWithOffset(1, format, args...)`

File: tests/e2e/tidbcluster/tidbcluster.go (hunk 17 above)
Comment (dragonly) [reply tests/e2e/tidbcluster/tidbcluster.go:2150]: What do you mean by fix?

File: tests/e2e/tidbcluster/tidbcluster.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply tests/e2e/tidbcluster/tidbcluster.go:2220]: just ignore this.

File: tests/e2e/tidbcluster/tidbcluster.go (hunk 17 above)
Comment (csuzhangxc) [reply tests/e2e/tidbcluster/tidbcluster.go:2121]: not `scale in` here, right?

Comment (dragonly) [reply tests/e2e/tidbcluster/tidbcluster.go]: fixed

**Review by csuzhangxc — APPROVED** [review]
LGTM

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Pull request merged**